### PR TITLE
MGMT 4531 refactor cnv operator

### PR DESCRIPTION
**Pull request by jakub-dzon**
Follow-up for https://github.com/openshift/assisted-service/pull/1329#discussion_r596001227 review comment.

### CODE REVIEW

File: internal/operators/cnv/cnv_operator.go [hunk 1]
```diff
@@ -80,77 +80,42 @@ func (o *operator) ValidateHost(ctx context.Context, cluster *common.Cluster, ho
...
 
-	if !hardware.IsVirtSupported(&inventory) {
+	if !virt.IsVirtSupported(&inventory) {
 		return api.ValidationResult{Status: api.Failure, ValidationId: o.GetClusterValidationID(), Reasons: []string{"CPU does not have virtualization support "}}, nil
 	}
 
 	// If the Role is set to Auto-assign for a host, it is not possible to determine whether the node will end up as a master or worker node.
 	if host.Role == models.HostRoleAutoAssign {
-		status := "All host roles must be assigned to enable OCS."
+		status := "All host roles must be assigned to enable CNV."
```
Comment (pkliczewski) [inline internal/operators/cnv/cnv_operator.go:89]: Good catch :)

File: internal/operators/cnv/cnv_operator.go [hunk 2]
```diff
@@ -80,77 +80,42 @@ func (o *operator) ValidateHost(ctx context.Context, cluster *common.Cluster, ho
...
-}
-
-// GetCPURequirementForMaster provides master CPU requirements for the operator
-func (o *operator) GetCPURequirementForMaster(_ context.Context, _ *common.Cluster) (int64, error) {
-	return masterCPU, nil
-}
+	requirements, err := o.GetHostRequirementsForRole(ctx, cluster, host.Role)
+	if err != nil {
+		o.log.Errorf("Failed to get host requirements for host with id %s", host.ID)
+		return api.ValidationResult{Status: api.Failure, ValidationId: o.GetClusterValidationID()}, err
```
Comment (pkliczewski) [inline internal/operators/cnv/cnv_operator.go:96]: What do you think about adding `Reasons` here?

File: internal/operators/cnv/cnv_operator.go [hunk 3]
```diff
@@ -160,3 +125,20 @@ func (o *operator) GetProperties() models.OperatorProperties {
...
+
+// GetHostRequirementsForRole provides operator's requirements towards host in a given role
+func (o *operator) GetHostRequirementsForRole(_ context.Context, _ *common.Cluster, role models.HostRole) (*models.ClusterHostRequirementsDetails, error) {
+	switch role {
+	case models.HostRoleMaster:
+		return &models.ClusterHostRequirementsDetails{
+			CPUCores: masterCPU,
+			RAMMib:   masterMemory,
+		}, nil
+	case models.HostRoleWorker, models.HostRoleAutoAssign:
```
Comment (pkliczewski) [inline internal/operators/cnv/cnv_operator.go:137]: I would not use `HostRoleAutoAssign` here. We are unable to validate when auto role is provided.

Comment (jakub-dzon) [reply internal/operators/cnv/cnv_operator.go]: Changed in #1329

File: internal/operators/cnv/cnv_operator.go (hunk 2 above)
Comment (jakub-dzon) [reply internal/operators/cnv/cnv_operator.go]: Done

**Comment by YuviGold** [comment]
Please rebase. It's difficult to review

**Comment by jakub-dzon** [comment]
Let's wait for #1329 to be merged
/hold

File: internal/operators/cnv/cnv_operator.go [hunk 4]
```diff
@@ -86,54 +86,30 @@ func (o *operator) ValidateHost(ctx context.Context, cluster *common.Cluster, ho
...
-
-// GetCPURequirementForMaster provides master CPU requirements for the operator
-func (o *operator) GetCPURequirementForMaster(_ context.Context, _ *common.Cluster) (int64, error) {
-	return masterCPU, nil
-}
+	requirements, err := o.GetHostRequirementsForRole(ctx, cluster, host.Role)
+	if err != nil {
+		message := fmt.Sprintf("Failed to get host requirements for host with id %s", host.ID)
+		o.log.Error(message)
+		return api.ValidationResult{Status: api.Failure, ValidationId: o.GetClusterValidationID(), Reasons: []string{message}}, err
```
Comment (pkliczewski) [inline internal/operators/cnv/cnv_operator.go:97]: What do you think about adding `err.Error()` to `Reasons`?

**Review by pkliczewski — COMMENTED** [review]
Looks good, small comment/suggestion

File: internal/operators/cnv/cnv_operator.go (hunk 4 above)
Comment (jakub-dzon) [reply internal/operators/cnv/cnv_operator.go:97]: OK, I think in this case it is safe to add it to `Reasons`. 

If there was a possibility to have the error coming from deep app internals or DB connection I would be against it, as there might be information in there that a user shouldn't see.

**Review by gamli75 — APPROVED** [review]

**Pull request merged**